### PR TITLE
Week 2 Solutions

### PR DESCRIPTION
**Pull request by talatmorshedtanbir**

### CODE REVIEW

File: week02/Implement strStr().cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
+// https://leetcode.com/problems/implement-strstr/
+// Time Complexity: O(N)
```
Comment (Sadman007) [inline week02/Implement strStr().cpp:2]: It is actually `O(N + M)`.

File: week02/Valid Parentheses.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            strStack.push(s[i]);
+            continue;
+        }
+
+        if(strStack.top()=='(' && s[i]==')')
+            strStack.pop();
+        else if(strStack.top()=='{' && s[i]=='}')
+            strStack.pop();
+        else if(strStack.top()=='['&& s[i]==']')
+            strStack.pop();
```
Comment (Sadman007) [inline week02/Valid Parentheses.cpp:21]: All of these `if-else`s could be replaced by a map.

**Review by Sadman007 — APPROVED** [review]

**Comment by Sadman007** [comment]
Hi @talatmorshedtanbir , nice work!

File: week02/Implement strStr().cpp (hunk 1 above)
Comment (talatmorshedtanbir) [reply week02/Implement strStr().cpp:2]: Missed it. Fixed

File: week02/Longest Substring Without Repeating Characters.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        for(int i=0; i<s.length(); i++) {
+            if(freq[s[i]] > start){
+                start = freq[s[i]];
+            }
+            freq[s[i]] = i;
+
+            maxLen = max(maxLen,i-start);
+        }
+
+        return maxLen;
```
[the remark this answers is not in the document]
Comment (talatmorshedtanbir) [reply week02/Longest Substring Without Repeating Characters.cpp:20]: Thanks vaiya

**Pull request merged**